android-components/components/feature/addons/src/main/java/mozilla/components/feature/addons/ui/AddonFilePicker.kt

```diff
@@ -0,0 +1,71 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.addons.ui
+
+import android.content.ActivityNotFoundException
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import androidx.activity.result.ActivityResultCaller
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import mozilla.components.feature.addons.Addon
+import mozilla.components.feature.addons.AddonManager
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.android.net.toFileUri
+
+/**
+ * // TODO: WRITE DOCS
+ */
+class AddonFilePicker(
+    val context: Context,
+    private val addonManager: AddonManager,
+) {
+
+    internal lateinit var activityLauncher: ActivityResultLauncher<Array<String>>
+    private val logger = Logger("AddonFilePicker")
+
+    fun launchFilePicker() {
+        try {
+            activityLauncher.launch(emptyArray())
+
+        } catch (e: ActivityNotFoundException) {
+            logger.error("Unable to find an app to select an XPI file", e)
+        }
+    }
+
+    fun registerResults(resultCaller: ActivityResultCaller) {
+        activityLauncher =
+            resultCaller.registerForActivityResult(AddonOpenDocument()) { uri ->
+                handleUriSelected(uri)
+            }
+    }
+
+    internal fun handleUriSelected(uri: Uri?) {
+        uri?.let {
+            // We want to verify that we are not exposing any private data
+            val fileUri = uri.toFileUri(context,"XIPs").toString()
+            val onSuccess: ((Addon) -> Unit) = {
+                logger.info("Add installed successfully")
+            }
+            val onError: ((String, Throwable) -> Unit) = { _, _ ->
+                logger.info("Unable to install add")
+            }
+            addonManager.installAddon(
+                Addon(id = "myAddon", downloadUrl = fileUri),
```

Comment:
ha, so we have to fake the `id` here?

Reply:
I see. That's interesting. Maybe we could sub-class `Addon` with a `LocalAddon` that would hardcode the `id` internally?

Reply:
See https://github.com/mozilla-mobile/firefox-android/pull/4568#discussion_r1403655894